Add support for additional, out-of-tree PAYG providers

wjt opened the pull request:
We need to be able to integrate with existing PAYG ecosystems, which come with their own flavours of unlock codes and their own non-libre verification libraries.

Outstanding:

* [x] Adjust packaging. In particular, we'll need a -dev package which installs the .pc file and a few headers. I think the shared library can just live in the main daemon package.
* [x] Fill in some docs which are currently of the form "..." or "document me".
* [x] Test this and an out-of-tree provider plugin in situ.
* [x] Update README
* [x] Maybe other things

https://phabricator.endlessm.com/T23606

pwithnall commented on libeos-payg/meson.build:
```diff
@@ -31,15 +37,40 @@ libeos_payg_resources = gnome.compile_resources(
+  ],
+  # According to http://mesonbuild.com/Pkgconfig-module.html#pkggenerate,
+  # libeos_payg's dependencies will automatically be added to Requires.private
+  # with Meson >= 0.45, but this doesn't seem true in my testing.
```
A Meson bug? Filed a bug report?

pwithnall commented on libeos-payg/multi-task.c:
```diff
@@ -92,6 +92,39 @@ epg_multi_task_return_boolean (GTask   *task,
+ * encountered. If an error has been encountered, or further operations are
+ * pending, @result is immediately destroyed with @result_destroy.
+ *
+ * One way to use this function is to use a %GPtrArray as @task's task_data.
```
Nitpick: s/`%`/`#`/ for types. `%` should be used for enum members.

pwithnall commented on libeos-payg/multi-task.c:
```diff
@@ -92,6 +92,39 @@ epg_multi_task_return_boolean (GTask   *task,
+ * One way to use this function is to use a %GPtrArray as @task's task_data.
+ * Each time a subtask completes, add its output to the array, and call
+ * `epg_multi_task_return_pointer (task, g_ptr_array_ref (array),
+ * g_ptr_array_unref);`.
```
Nitpick: Missing `Since: …` line

pwithnall commented on libeos-payg/multi-task.h:
```diff
@@ -30,6 +30,9 @@ void epg_multi_task_attach         (GTask         *task,
 void epg_multi_task_increment      (GTask         *task);
 void epg_multi_task_return_boolean (GTask         *task,
                                     gboolean       result);
+void epg_multi_task_return_pointer (GTask         *task,
```
Note: As discussed in person in London, it would be good if `EpgMultiTask` could be submitted upstream to GLib. However, that’s definitely not a prerequisite for this PR. I’m not going to do API review on it here, since the API seems appropriate for its use in this PR.

Upstream-wise, I think the API we discussed was to have the `MultiTask` expose all the return values from all its sub-tasks, and allow the caller to implement their own result accumulation; rather than trying to provide some generic mechanism for result accumulation.

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+/**
+ * EpgProviderLoader:
+ *
+ * TODO: write me
```
Note: A wild `TODO` comment! :lion:

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+  if (self->module_dir == NULL)
+    self->module_dir = g_getenv ("EPG_MODULE_DIR");
+
+  if (self->module_dir == NULL)
```
Might want to squash `EPG_MODULE_DIR=""` down to `NULL` here and ignore it?

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+ * epg_provider_loader_load_async:
+ * @self: a %EpgProviderLoader
+ * @enabled: ...
+ * ...
```
A wild `...` comment! :tiger:

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+      if (ret)
+        {
+          /* Should not happen: peas_engine_load_plugin() should cause the
+           * PeasPluginInfo's error to be set.
```
If this ‘should not happen’, perhaps it should be an assertion?

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+                                            NULL);
+  if (extension == NULL)
+    {
+      /* Should not happen: we just checked that plugin_info can do this. */
```
Again, should this be an assertion?

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+{
+  g_autoptr(EpgProvider) provider = EPG_PROVIDER (source_object);
+  g_autoptr(GTask) task = G_TASK (user_data);
+  g_autoptr(GError) error = NULL;
```
Nitpick: I prefer the name `local_error` to be used for local `GError*` variables, to distinguish it from `GError **error` arguments.

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+
+/**
+ * epg_provider_loader_load_finish:
+ *
```
Needs some documentation.

pwithnall commented on libeos-payg/provider-loader.c:
```diff
@@ -0,0 +1,310 @@
+  /* At present, we always return success, even if we found >= 1 plugin but all
+   * failed to load/initialize for some reason. Should we return an error in
+   * that case? If so: if two plugins are found and both fail to load, which
+   * error should we return?
```
I guess it depends where you want the policy of “exactly one provider must be loaded” to be implemented. Given that the code to load the internal fallback provider is not in the `ProviderLoader`, that suggests to me that the `ProviderLoader` should just return the array of providers it to loaded (which may be empty), and let the caller figure out what to do next (which might include erroring).

That appears to be what you’ve done anyway. Maybe just turn this comment into an explanation of that.

pwithnall commented on meson.build:
```diff
@@ -22,6 +22,8 @@ localedir = join_paths(prefix, get_option('localedir'))
 includedir = join_paths(prefix, get_option('includedir'))
 localstatedir = join_paths(prefix, get_option('localstatedir'))
 sysconfdir = join_paths(prefix, get_option('sysconfdir'))
+pkglibdir = join_paths(libdir, meson.project_name()) # TODO: -1 suffix? soname?
```
Any suffix would be for the purposes of parallel-installability, and hence would have to include the major version of the project, and the major version of the plugin API. Those versions might be linked so that only one version number needs to be included.

I would recommend adding in a major version number suffix here, yes.

pwithnall commented on libeos-payg/service.c:
```diff
@@ -125,22 +126,90 @@ epg_service_get_main_option_entries (GssService *service)
+  else
+    {
+      EpgProvider *provider = NULL;
+      guint i;
```
Nitpick: `gsize` for array iteration please

wjt replied:
I think it is correct to match the type of `GPtrArray.len`:

```c
struct GPtrArray {
  gpointer *pdata;
  guint	    len;
};
```

pwithnall replied:
(If I could go back in time and change `GPtrArray.len` to be a `gsize`, I would. Then we could legitimately address every byte of address space in a `GPtrArray` and the world would somehow be a better place.)

wjt replied:
And we could address more than 80 GiB of instances of EpgProvider implementations :-)

pwithnall commented on libeos-payg/service.c:
```diff
@@ -125,22 +126,90 @@ epg_service_get_main_option_entries (GssService *service)
+      for (i = 0; i < providers->len; i++)
+        {
+          EpgProvider *spare = EPG_PROVIDER (g_ptr_array_index (providers, i));
+          epg_provider_shutdown_async (spare, NULL, NULL, NULL);
```
I’m a little wary of this hanging async operation. There’s nothing which guarantees it’s completed before the `EpgService` is destroyed. Although unlikely, historically this kind of thing has been a source of bugs.

wjt replied:
So what would you have this do – wait for all shutdowns to finish before pressing on with the main event?

pwithnall commented on libeos-payg/service.c:
```diff
@@ -184,15 +253,35 @@ manager_new_cb (GObject      *source_object,
+ * @self:
+ * @task: (transfer none):
+ * @provider: (transfer full):
+ *
```
Needs documentation.

wjt replied:
Does this internal function need any more than this? I only wrote a docstring at all to have a place to put the transfer annotations. I am happy to just delete it.

pwithnall commented on libeos-payg/service.c:
```diff
@@ -184,15 +253,35 @@ manager_new_cb (GObject      *source_object,
+{
+  g_autoptr(GError) local_error = NULL;
+
+  self->provider = provider;
```
Nitpick: Might make things a little less likely to be broken in future to add a comment like `/* Take ownership */` above this assignment.

pwithnall commented on libeos-payg/service.c:
```diff
@@ -184,15 +253,35 @@ manager_new_cb (GObject      *source_object,
       return;
     }
 
+  epg_service_startup_complete (self, task, provider);
```
I’d do `g_steal_pointer (&provider)` instead of just passing `provider` in order to make it explicit that this takes ownership of it. Otherwise this could look like a memory leak.

pwithnall commented on libeos-payg/service.c:
```diff
@@ -172,8 +241,8 @@ epg_service_startup_async (GssService          *service,
 
-  epg_manager_new (enabled, NULL, NULL,
-                   cancellable,
+  GCancellable *cancellable = g_task_get_cancellable (task);
```
Maybe worth adding a comment here like `/* No provider plugin could be initialised. Fall back to #EpgManager. */`?

pwithnall replied:
Or maybe a bit further up by the `g_debug()` that kicks this all off.

pwithnall commented on libeos-payg/service.c:
```diff
@@ -125,22 +126,90 @@ epg_service_get_main_option_entries (GssService *service)
+
+      if (provider != NULL)
+        {
+          epg_service_startup_complete (self, task, provider);
```
I’d do `g_steal_pointer (&provider)` instead of just passing `provider` in order to make it explicit that this takes ownership of it. Otherwise this could look like a memory leak.

wjt commented on libeos-payg/tests/manager.c:
```diff
@@ -0,0 +1,348 @@
+      /* The manager internally performs asynchronous saves in a GTask thread.
+       * There's currently no way to wait for them to finish. If it happens that
+       * one of the files gets recreated after being removed above but before
+       * we get here, removing the directory will fail. Ignore the error.
```
One of the later patches does add such a shutdown API, but I didn't arrange for the EpgManager implementation of it to wait for any other save operations to complete – maybe I should.

wjt replied:
Addressed by finding a GLib bug, working around it, and fixing this problem

wjt commented:
test this please

wjt commented:
test this please

wjt commented on libeos-payg/tests/manager.c:
```diff
@@ -0,0 +1,348 @@
+  g_assert_nonnull (fixture->tmp_path);
+
+  fixture->tmp_dir = g_file_new_for_path (fixture->tmp_path);
+  g_assert_nonnull (fixture->tmp_dir);
```
Removed.

wjt commented on libeos-payg/tests/manager.c:
```diff
@@ -176,6 +184,112 @@ test_manager_load_empty (Fixture *fixture,
+  g_clear_error (&error);
+
+  /* Ideally, we would attempt to clean away the offending illegible 'file',
+   * but that is not currently the case.
```
It's just not implemented.

wjt commented on libeos-payg/tests/manager.c:
```diff
@@ -0,0 +1,348 @@
+  g_assert_true (ret);
+}
+
+static void
```
I wrote some comments.

wjt commented on libeos-payg/manager.c:
```diff
@@ -1070,28 +1050,21 @@ file_load_delete_cb (GObject      *source_object,
-  /* Decrement the pending operation count. */
-  guint operation_count = GPOINTER_TO_UINT (g_task_get_task_data (task));
-  g_task_set_task_data (task, GUINT_TO_POINTER (--operation_count), NULL);
+  g_autoptr(GError) error = NULL;
```
Pushed a fixup to undo this.

wjt commented on libeos-payg/provider.c:
```diff
@@ -0,0 +1,260 @@
+{
+  g_return_val_if_fail (EPG_IS_PROVIDER (self), 0);
+
+  return EPG_PROVIDER_GET_IFACE (self)->get_expiry_time (self);
```
I don't know which interface init function you mean: the only one needed by G_DEFINE_INTERFACE is epg_provider_default_init() and I think that's called on the interface itself, not for implementations of it. I added assertions to each wrapper function instead.

pwithnall replied:
I was misremembering `g_type_add_interface_check()`, but it should do what you want. Asserting in the wrapper functions works almost as well though.

wjt replied:
This function is … not widely used! https://github.com/search?q=org%3AGNOME+g_type_add_interface_check&type=Code

wjt commented:
bot, build

wjt commented:
test this please

wjt commented on libeos-payg/manager.c:
```diff
@@ -1064,7 +1072,7 @@ static void file_save_delete_cb (GObject      *source_object,
                                  gpointer      user_data);
 
-void
+static void
```
Yes, they should have been part of `Extract an EpgProvider interface from EpgManager`. I can only realistically fix this by going back and amending the commits manually at merge time.

pwithnall replied:
I’m sorry to have brought it up.

pwithnall commented on libeos-payg/multi-task.c:
```diff
@@ -39,12 +39,27 @@ epg_multi_task_set (GTask *task,
+ * decremented with epg_multi_task_return_boolean(),
+ * epg_multi_task_return_pointer(), and epg_multi_task_return_error().
+ *
+ * The number of pending operations is stored by occult means, leaving the
```
More documentation should be like this. ⛧

wjt replied:
Credit where credit's due: I stole this phrase from a [New Yorker review of Black Mirror](https://www.newyorker.com/magazine/2015/01/05/button-pusher) on the occasion of its legal release in the US:

> For a full day after watching the first episode (which I obtained through occult means, before Netﬂix made the show available to U.S. viewers), I felt disoriented, dropped on a new planet.

pwithnall commented on libeos-payg/tests/manager.c:
```diff
@@ -408,6 +408,62 @@ test_manager_add_save_reload (Fixture *fixture,
+ *
+ * Tests that applying an infinite code sets the expiry time to infinitely far
+ * in the future, and that this survives both a restart and adding a
+ * non-infinite code afterwards.
```
Nice.

pwithnall commented on meson.build:
```diff
@@ -79,7 +79,6 @@ test_args = [
   '-Wmissing-prototypes',
   '-Wnested-externs',
   '-Wno-error=cpp',
-  '-Wno-discarded-qualifiers',
```
Maybe explicitly switch this to `-Wdiscarded-qualifiers` to make it clear that it’s a warning we care about?

pwithnall commented on libeos-payg/manager.c:
```diff
@@ -796,8 +814,15 @@ epg_manager_add_code (EpgProvider   *provider,
+   * FIXME: pass self->cancellable; see comment in
+   * epg_manager_shutdown_async().
+   */
+  g_assert_cmpuint (self->pending_internal_save_state_calls, <, G_MAXUINT64);
```
Use `g_assert()`, rather than `g_assert_*()`, in non-test code. `g_assert_*()` log through the test logging code, and are affected by various test settings. `g_assert()` is disableable by compiling with `G_DISABLE_ASSERT`, which people typically want for higher performance production code. It also isn’t affected by any of the test settings.

pwithnall commented on libeos-payg/manager.c:
```diff
@@ -822,12 +850,45 @@ epg_manager_clear_code (EpgProvider  *provider,
+   * FIXME: pass self->cancellable; see comment in
+   * epg_manager_shutdown_async().
+   */
+  g_assert_cmpuint (self->pending_internal_save_state_calls, <, G_MAXUINT64);
```
Same here: use `g_assert()`.

pwithnall commented on libeos-payg/manager.c:
```diff
@@ -822,12 +850,45 @@ epg_manager_clear_code (EpgProvider  *provider,
+  g_return_if_fail (self->pending_internal_save_state_calls > 0);
+  self->pending_internal_save_state_calls--;
+
+  if (self->pending_internal_save_state_calls == 0 &&
```
This block deserves a bit of a comment, I think, since it’s effectively kicking off another save operation directly after a previous one has (potentially successfully) finished.

wjt replied:
Hmm. Actually, this is only needed because if there was more than one pending call, it's not guaranteed that the later one "wins". Perhaps I should change all this logic to allow:

* at most 1 save operation in flight (with 1 or more tasks chained to it)
* a possibly-empty queue of requests to be serviced (ideally all at once) when the in-flight request completes

It seems like a lot of :keyboard: but it would have the advantage of being more robust in the face of unexpected power loss. I only have limited appetite for caring about this, though, because in ordinary use it's not possible to have >1 save at a time (since the UI only allows you to enter a single code, only when the machine is locked).

wjt replied:
The less laborious option is to make shutdown_async something like:

```
if no saves in flight:
    return True # nothing to do
else:
    wait for last save to finish, propagate its result
```

This changes the behaviour in the case where writing back the state would fail, but the state hasn't changed so there's no need to write it back. It also ignores the problem that if two saves are in flight it's not defined which would finish first. But it is a lot easier…

wjt replied:
I implemented:

```
if no saves in flight:
   save once more
else:
   wait for last to finish & propagate result
```

pwithnall commented on libeos-payg/provider.c:
```diff
@@ -90,20 +90,25 @@ epg_provider_default_init (EpgProviderInterface *iface)
-   * - `"*00000000000000#"`: codes begin with `*`, followed by 14 digits,
-   *   followed by `#`
+   * - `^\d{8}$`: codes are 8 arbitrary digits, such as `"12345678"` or
+   *   `"१२३४५६७८"` (use `[0-9]` if you mean ASCII digits)
```
I didn’t think it was possible to get this obscure. You win points.

pwithnall commented on libeos-payg/tests/manager.c:
```diff
@@ -617,6 +705,30 @@ main (int    argc,
+  const char *non_matches[] = {
+      "\n", "a", "1a", "123\n", "a123",
+      /* '6' in various non-ASCII forms */
+      "६", "೬", "𝟨", "六",
```
I can see you’ve had a lot of fun here. This review has been a rollercoaster of emotion.

pwithnall commented on libeos-payg/tests/manager.c:
```diff
@@ -595,6 +665,24 @@ test_manager_error_rate_limit (Fixture *fixture,
+static void
+add_many (const char  *path_base,
+          void       (*func) (Fixture *, gconstpointer),
+          const char **data)
```
Are you a practizer of the occult art of functional programming, perchance?

pwithnall commented on libeos-payg/tests/plugins/meson.build:
```diff
@@ -0,0 +1,42 @@
+  configure_file(
+    input: 'epg-test-provider.plugin.in',
+    output: 'epg-test-provider-' + suffix + '.plugin',
+    configuration: plugin_conf
```
Nitpick: I’d add a trailing comma here for future-proofness.

pwithnall commented on libeos-payg/tests/plugins/test-provider.c:
```diff
@@ -0,0 +1,319 @@
+  PROP_ENABLED,
+  PROP_RATE_LIMIT_END_TIME,
+  PROP_CODE_FORMAT,
+} EpgProviderAngazaProperty;
```
I suspect this should be `EpgTestProviderProperty`.

wjt replied:
I think you're right!

pwithnall commented on libeos-payg/tests/plugins/test-provider.c:
```diff
@@ -0,0 +1,319 @@
+
+static void
+epg_test_provider_async_initable_iface_init (gpointer g_iface,
+                                               gpointer iface_data)
```
Nitpick: Indentation problem (and in a few other places in this file).

wjt commented:
Rebased and squashed. Many conflicts but there was no diff to the pre-squashed version once complete, and `ninja test` passed after every conflict resolution.
